sklearnex/neighbors/common.py

```diff
@@ -15,15 +15,118 @@
+from daal4py.sklearn._utils import PatchingConditionsChain, sklearn_check_version
+from onedal.datatypes import _check_array, _num_features, _num_samples
+
+import numpy as np
+from scipy import sparse as sp
+import warnings
+
+from sklearn.neighbors._base import VALID_METRICS
 from sklearn.neighbors._base import NeighborsBase as sklearn_NeighborsBase
 from sklearn.neighbors._ball_tree import BallTree
 from sklearn.neighbors._kd_tree import KDTree
-import numpy as np
-from scipy import sparse as sp
 
 
 class KNeighborsDispatchingBase:
+    def _fit_validation(self, X, y=None):
+        if sklearn_check_version("1.2"):
+            self._validate_params()
+        if sklearn_check_version("1.0"):
+            self._check_feature_names(X, reset=True)
+        if self.metric_params is not None and 'p' in self.metric_params:
+            if self.p is not None:
+                warnings.warn("Parameter p is found in metric_params. "
+                              "The corresponding parameter from __init__ "
+                              "is ignored.", SyntaxWarning, stacklevel=2)
+            self.effective_metric_params_ = self.metric_params.copy()
+            effective_p = self.metric_params["p"]
+        else:
+            self.effective_metric_params_ = {}
+            effective_p = self.p
+
+        self.effective_metric_params_["p"] = effective_p
+        self.effective_metric_ = self.metric
+        # For minkowski distance, use more efficient methods where available
+        if self.metric == "minkowski":
+            p = self.effective_metric_params_["p"]
+            if p == 1:
+                self.effective_metric_ = "manhattan"
+            elif p == 2:
+                self.effective_metric_ = "euclidean"
+            elif p == np.inf:
+                self.effective_metric_ = "chebyshev"
+
+        if not isinstance(X, (KDTree, BallTree, sklearn_NeighborsBase)):
+            self._fit_X = _check_array(
+                X, dtype=[np.float64, np.float32], accept_sparse=True)
+            self.n_samples_fit_ = _num_samples(self._fit_X)
+            self.n_features_in_ = _num_features(self._fit_X)
+
+            if self.algorithm == "auto":
+                # A tree approach is better for small number of neighbors or small
+                # number of features, with KDTree generally faster when available
+                is_n_neighbors_valid_for_brute = self.n_neighbors is not None and \
+                    self.n_neighbors >= self._fit_X.shape[0] // 2
+                if self._fit_X.shape[1] > 15 or is_n_neighbors_valid_for_brute:
+                    self._fit_method = "brute"
+                else:
+                    if self.effective_metric_ in VALID_METRICS["kd_tree"]:
+                        self._fit_method = "kd_tree"
+                    elif callable(self.effective_metric_) or \
+                        self.effective_metric_ in \
+                            VALID_METRICS["ball_tree"]:
```

Comment:
What about completely removing `ball_tree` from the default branch? It is not implemented on the oneDAL side.

Another question - can we disable `kd_tree` for GPUs?

Reply:
`self._fit_method="ball_tree"` is required later to decide if oneDAL patching is applicable. `kd_tree` on GPU is disabled - see `_onedal_supported` function